# I. Clean the files and combine them into one final DataFrame.

This dataframe should have the following columns:
Hero (Just the name of the Hero)
Publisher
Gender
Eye color
Race
Hair color
Height (numeric)
Skin color
Alignment
Weight (numeric)
Plus, one-hot-encoded columns for every power that appears in the dataset. E.g.:
Agility
Flight
Superspeed
etc.

# Imports and Libraries

In [1]:
## Standard Imports
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
## Importing the OS and JSON Modules
import os,json

# Cleaning of dataset Number 1

In [2]:
dfsh = pd.read_csv('C:/Users/16024/Desktop/superhero_info - superhero_info.csv')
dfsh.head()

Hero|Publisher Gender               Race Alignment  \
0          A-Bomb|Marvel Comics   Male              Human      good   
1  Abe Sapien|Dark Horse Comics   Male      Icthyo Sapien      good   
2            Abin Sur|DC Comics   Male            Ungaran      good   
3     Abomination|Marvel Comics   Male  Human / Radiation       bad   
4   Absorbing Man|Marvel Comics   Male              Human       bad   

  Hair color Eye color Skin color  \
0    No Hair    yellow    Unknown   
1    No Hair      blue       blue   
2    No Hair      blue        red   
3    No Hair     green    Unknown   
4    No Hair      blue    Unknown   

                                   Measurements  
0  {'Height': '203.0 cm', 'Weight': '441.0 kg'}  
1   {'Height': '191.0 cm', 'Weight': '65.0 kg'}  
2   {'Height': '185.0 cm', 'Weight': '90.0 kg'}  
3  {'Height': '203.0 cm', 'Weight': '441.0 kg'}  
4  {'Height': '193.0 cm', 'Weight': '122.0 kg'}

In [3]:
dfsh.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 463 entries, 0 to 462
Data columns (total 8 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   Hero|Publisher  463 non-null    object
 1   Gender          463 non-null    object
 2   Race            463 non-null    object
 3   Alignment       463 non-null    object
 4   Hair color      463 non-null    object
 5   Eye color       463 non-null    object
 6   Skin color      463 non-null    object
 7   Measurements    463 non-null    object
dtypes: object(8)
memory usage: 29.1+ KB


In [4]:
dfsh.shape

(463, 8)

In [5]:
dfsh.duplicated().sum().sum()

0

# Cleaning of DataSet Number 2

In [6]:
dfsp = pd.read_csv('C:/Users/16024/Desktop/superhero_powers - superhero_powers (1).csv')
dfsp.head()

hero_names                                             Powers
0      3-D Man         Agility,Super Strength,Stamina,Super Speed
1       A-Bomb  Accelerated Healing,Durability,Longevity,Super...
2   Abe Sapien  Agility,Accelerated Healing,Cold Resistance,Du...
3     Abin Sur                                 Lantern Power Ring
4  Abomination  Accelerated Healing,Intelligence,Super Strengt...

In [7]:
dfsp.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 667 entries, 0 to 666
Data columns (total 2 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   hero_names  667 non-null    object
 1   Powers      667 non-null    object
dtypes: object(2)
memory usage: 10.5+ KB


In [8]:
dfsp.shape

(667, 2)

In [9]:
dfsp.duplicated().sum().sum()

0

# Hero and Publisher

In [10]:
dfsh[["Hero", "Publisher"]] = dfsh["Hero|Publisher"].str.split('|',expand=True)
dfsh.head()

Hero|Publisher Gender               Race Alignment  \
0          A-Bomb|Marvel Comics   Male              Human      good   
1  Abe Sapien|Dark Horse Comics   Male      Icthyo Sapien      good   
2            Abin Sur|DC Comics   Male            Ungaran      good   
3     Abomination|Marvel Comics   Male  Human / Radiation       bad   
4   Absorbing Man|Marvel Comics   Male              Human       bad   

  Hair color Eye color Skin color  \
0    No Hair    yellow    Unknown   
1    No Hair      blue       blue   
2    No Hair      blue        red   
3    No Hair     green    Unknown   
4    No Hair      blue    Unknown   

                                   Measurements           Hero  \
0  {'Height': '203.0 cm', 'Weight': '441.0 kg'}         A-Bomb   
1   {'Height': '191.0 cm', 'Weight': '65.0 kg'}     Abe Sapien   
2   {'Height': '185.0 cm', 'Weight': '90.0 kg'}       Abin Sur   
3  {'Height': '203.0 cm', 'Weight': '441.0 kg'}    Abomination   
4  {'Height': '193.0 cm', 'Weight': '122.0 kg'}  Absorbing Man   

           Publisher  
0      Marvel Comics  
1  Dark Horse Comics  
2          DC Comics  
3      Marvel Comics  
4      Marvel Comics

In [11]:
# drop original 'Hero|Publisher' column
dfsh = dfsh.drop(columns = ["Hero|Publisher"])

# check
dfsh.head(2)

Gender           Race Alignment Hair color Eye color Skin color  \
0   Male          Human      good    No Hair    yellow    Unknown   
1   Male  Icthyo Sapien      good    No Hair      blue       blue   

                                   Measurements        Hero          Publisher  
0  {'Height': '203.0 cm', 'Weight': '441.0 kg'}      A-Bomb      Marvel Comics  
1   {'Height': '191.0 cm', 'Weight': '65.0 kg'}  Abe Sapien  Dark Horse Comics

In [12]:
# create an order that makes more sense
dfsh = dfsh[["Publisher","Hero","Gender","Race","Alignment","Hair color","Eye color","Skin color", "Measurements"]]
dfsh.head()

Publisher           Hero Gender               Race Alignment  \
0      Marvel Comics         A-Bomb   Male              Human      good   
1  Dark Horse Comics     Abe Sapien   Male      Icthyo Sapien      good   
2          DC Comics       Abin Sur   Male            Ungaran      good   
3      Marvel Comics    Abomination   Male  Human / Radiation       bad   
4      Marvel Comics  Absorbing Man   Male              Human       bad   

  Hair color Eye color Skin color  \
0    No Hair    yellow    Unknown   
1    No Hair      blue       blue   
2    No Hair      blue        red   
3    No Hair     green    Unknown   
4    No Hair      blue    Unknown   

                                   Measurements  
0  {'Height': '203.0 cm', 'Weight': '441.0 kg'}  
1   {'Height': '191.0 cm', 'Weight': '65.0 kg'}  
2   {'Height': '185.0 cm', 'Weight': '90.0 kg'}  
3  {'Height': '203.0 cm', 'Weight': '441.0 kg'}  
4  {'Height': '193.0 cm', 'Weight': '122.0 kg'}

# Heights and Weights

In [13]:
# check type of 'Measurements' column
measurement = dfsh.loc[0, 'Measurements']
print(type(measurement))
print(measurement)

<class 'str'>
{'Height': '203.0 cm', 'Weight': '441.0 kg'}


In [14]:
# convert strings in dictionary to using double instead of
# single quotes for json.loads compatibility
dfsh['Measurements'] = dfsh['Measurements'].str.replace("'", '"')

# check
measurement = dfsh.loc[0, 'Measurements']
print(measurement)

{"Height": "203.0 cm", "Weight": "441.0 kg"}


In [15]:
# use .apply with json.loads to convert strings to dicts
dfsh['Measurements'] = dfsh['Measurements'].apply(json.loads)

# check
measurement = dfsh.loc[0, 'Measurements']
print(type(measurement))

<class 'dict'>


In [16]:
# unpack 'Measurements' column (dictionaries) into separate
# columns for each dictionary item
height_weight = dfsh['Measurements'].apply(pd.Series)

# check
height_weight

Height    Weight
0    203.0 cm  441.0 kg
1    191.0 cm   65.0 kg
2    185.0 cm   90.0 kg
3    203.0 cm  441.0 kg
4    193.0 cm  122.0 kg
..        ...       ...
458  183.0 cm   83.0 kg
459  165.0 cm   52.0 kg
460   66.0 cm   17.0 kg
461  170.0 cm   57.0 kg
462  185.0 cm   81.0 kg

[463 rows x 2 columns]

In [17]:
# concatenate new columns to original df
dfsh = pd.concat((dfsh, height_weight), axis = 1)

# check
dfsh.head(2)

Publisher        Hero Gender           Race Alignment Hair color  \
0      Marvel Comics      A-Bomb   Male          Human      good    No Hair   
1  Dark Horse Comics  Abe Sapien   Male  Icthyo Sapien      good    No Hair   

  Eye color Skin color                                  Measurements  \
0    yellow    Unknown  {'Height': '203.0 cm', 'Weight': '441.0 kg'}   
1      blue       blue   {'Height': '191.0 cm', 'Weight': '65.0 kg'}   

     Height    Weight  
0  203.0 cm  441.0 kg  
1  191.0 cm   65.0 kg

In [18]:
# drop original column
dfsh = dfsh.drop(columns = ['Measurements'])

# check
dfsh.head(2)

Publisher        Hero Gender           Race Alignment Hair color  \
0      Marvel Comics      A-Bomb   Male          Human      good    No Hair   
1  Dark Horse Comics  Abe Sapien   Male  Icthyo Sapien      good    No Hair   

  Eye color Skin color    Height    Weight  
0    yellow    Unknown  203.0 cm  441.0 kg  
1      blue       blue  191.0 cm   65.0 kg

In [19]:
# change 'Height' and 'Weight' column names
dfsh.rename(columns = {'Height': 'Height (cm)',
                      'Weight': 'Weight (kg)'}, 
            inplace = True)

# check
dfsh.head(2)

Publisher        Hero Gender           Race Alignment Hair color  \
0      Marvel Comics      A-Bomb   Male          Human      good    No Hair   
1  Dark Horse Comics  Abe Sapien   Male  Icthyo Sapien      good    No Hair   

  Eye color Skin color Height (cm) Weight (kg)  
0    yellow    Unknown    203.0 cm    441.0 kg  
1      blue       blue    191.0 cm     65.0 kg

In [20]:
# take off 'cm' and 'kg' from cell values in 'Height' and 'Weight'
to_replace = [' cm', ' kg']

for char in to_replace:
    dfsh['Height (cm)'] = dfsh['Height (cm)'].str.replace(char, 
                                                          '', 
                                                          regex = False)
    dfsh['Weight (kg)'] = dfsh['Weight (kg)'].str.replace(char, 
                                                          '', 
                                                          regex = False)
    
# check
dfsh[['Height (cm)', 'Weight (kg)']].head(2)

Height (cm) Weight (kg)
0       203.0       441.0
1       191.0        65.0

In [21]:
# convert 'Height (cm)' and 'Weight (kg)' columns from object-
# types to numeric

dfsh['Height (cm)'] = dfsh['Height (cm)'].astype(float)
dfsh['Weight (kg)'] = dfsh['Weight (kg)'].astype(float)

# check
dfsh.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 463 entries, 0 to 462
Data columns (total 10 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Publisher    463 non-null    object 
 1   Hero         463 non-null    object 
 2   Gender       463 non-null    object 
 3   Race         463 non-null    object 
 4   Alignment    463 non-null    object 
 5   Hair color   463 non-null    object 
 6   Eye color    463 non-null    object 
 7   Skin color   463 non-null    object 
 8   Height (cm)  463 non-null    float64
 9   Weight (kg)  463 non-null    float64
dtypes: float64(2), object(8)
memory usage: 36.3+ KB


# Powers

In [22]:
# check type of powers['Powers'] column
power = dfsp.loc[0, 'Powers']
print(type(power))
print(power)

<class 'str'>
Agility,Super Strength,Stamina,Super Speed


In [23]:
# split string on comma
dfsp['powers_split'] = dfsp['Powers'].str.split(',')

# check
dfsp[['Powers', 'powers_split']].head(2)

Powers  \
0         Agility,Super Strength,Stamina,Super Speed   
1  Accelerated Healing,Durability,Longevity,Super...   

                                        powers_split  
0    [Agility, Super Strength, Stamina, Super Speed]  
1  [Accelerated Healing, Durability, Longevity, S...

In [24]:
exploded = dfsp.explode('powers_split')
exploded[['hero_names', 'Powers', 'powers_split']].head(5)

hero_names                                             Powers  \
0    3-D Man         Agility,Super Strength,Stamina,Super Speed   
0    3-D Man         Agility,Super Strength,Stamina,Super Speed   
0    3-D Man         Agility,Super Strength,Stamina,Super Speed   
0    3-D Man         Agility,Super Strength,Stamina,Super Speed   
1     A-Bomb  Accelerated Healing,Durability,Longevity,Super...   

          powers_split  
0              Agility  
0       Super Strength  
0              Stamina  
0          Super Speed  
1  Accelerated Healing

In [25]:
# save unique values from exploded column
cols_to_make = exploded['powers_split'].dropna().unique()

# check
cols_to_make

array(['Agility', 'Super Strength', 'Stamina', 'Super Speed',
       'Accelerated Healing', 'Durability', 'Longevity', 'Camouflage',
       'Self-Sustenance', 'Cold Resistance', 'Underwater breathing',
       'Marksmanship', 'Weapons Master', 'Intelligence', 'Telepathy',
       'Immortality', 'Reflexes', 'Enhanced Sight', 'Sub-Mariner',
       'Lantern Power Ring', 'Invulnerability', 'Animation',
       'Super Breath', 'Dimensional Awareness', 'Flight', 'Size Changing',
       'Teleportation', 'Magic', 'Dimensional Travel',
       'Molecular Manipulation', 'Energy Manipulation', 'Power Cosmic',
       'Energy Absorption', 'Elemental Transmogrification',
       'Fire Resistance', 'Natural Armor', 'Heat Resistance',
       'Matter Absorption', 'Regeneration', 'Stealth', 'Power Suit',
       'Energy Blasts', 'Energy Beams', 'Heat Generation', 'Danger Sense',
       'Phasing', 'Force Fields', 'Hypnokinesis', 'Invisibility',
       'Enhanced Senses', 'Jump', 'Shapeshifting', 'Elasticity',
 

In [27]:
# create new column for each unique value in list
# populate cells with True or False

frames = [dfsp]

for col in cols_to_make:
    frame = pd.DataFrame()
    frame[col] = dfsp['Powers'].str.contains(col)
    frames.append(frame)
    merged_powers = pd.concat(frames, axis = 1)
    
# check
merged_powers.head()

hero_names                                             Powers  \
0      3-D Man         Agility,Super Strength,Stamina,Super Speed   
1       A-Bomb  Accelerated Healing,Durability,Longevity,Super...   
2   Abe Sapien  Agility,Accelerated Healing,Cold Resistance,Du...   
3     Abin Sur                                 Lantern Power Ring   
4  Abomination  Accelerated Healing,Intelligence,Super Strengt...   

                                        powers_split  Agility  Super Strength  \
0    [Agility, Super Strength, Stamina, Super Speed]     True            True   
1  [Accelerated Healing, Durability, Longevity, S...    False            True   
2  [Agility, Accelerated Healing, Cold Resistance...     True            True   
3                               [Lantern Power Ring]    False           False   
4  [Accelerated Healing, Intelligence, Super Stre...    False            True   

   Stamina  Super Speed  Accelerated Healing  Durability  Longevity  ...  \
0     True         True                False       False      False  ...   
1     True        False                 True        True       True  ...   
2     True        False                 True        True       True  ...   
3    False        False                False       False      False  ...   
4     True         True                 True       False      False  ...   

   Weather Control  Omnipresent  Omniscient  Hair Manipulation  Nova Force  \
0            False        False       False              False       False   
1            False        False       False              False       False   
2            False        False       False              False       False   
3            False        False       False              False       False   
4            False        False       False              False       False   

   Odin Force  Phoenix Force  Intuitive aptitude  Melting  Changing Armor  
0       False          False               False    False           False  
1       False          False               False    False           False  
2       False          False               False    False           False  
3       False          False               False    False           False  
4       False          False               False    False           False  

[5 rows x 170 columns]

In [28]:
# drop 'Powers' and 'powers_split' columns
merged_powers.drop(columns = ['Powers', 'powers_split'], inplace = True)

# check
merged_powers.head()

hero_names  Agility  Super Strength  Stamina  Super Speed  \
0      3-D Man     True            True     True         True   
1       A-Bomb    False            True     True        False   
2   Abe Sapien     True            True     True        False   
3     Abin Sur    False           False    False        False   
4  Abomination    False            True     True         True   

   Accelerated Healing  Durability  Longevity  Camouflage  Self-Sustenance  \
0                False       False      False       False            False   
1                 True        True       True        True             True   
2                 True        True       True       False            False   
3                False       False      False       False            False   
4                 True       False      False       False            False   

   ...  Weather Control  Omnipresent  Omniscient  Hair Manipulation  \
0  ...            False        False       False              False   
1  ...            False        False       False              False   
2  ...            False        False       False              False   
3  ...            False        False       False              False   
4  ...            False        False       False              False   

   Nova Force  Odin Force  Phoenix Force  Intuitive aptitude  Melting  \
0       False       False          False               False    False   
1       False       False          False               False    False   
2       False       False          False               False    False   
3       False       False          False               False    False   
4       False       False          False               False    False   

   Changing Armor  
0           False  
1           False  
2           False  
3           False  
4           False  

[5 rows x 168 columns]

In [29]:
# change 'hero_names' column name to match info['Hero']
merged_powers.rename(columns = {'hero_names': 'Hero'}, 
                     inplace = True)

# check
merged_powers.head(2)

Hero  Agility  Super Strength  Stamina  Super Speed  \
0  3-D Man     True            True     True         True   
1   A-Bomb    False            True     True        False   

   Accelerated Healing  Durability  Longevity  Camouflage  Self-Sustenance  \
0                False       False      False       False            False   
1                 True        True       True        True             True   

   ...  Weather Control  Omnipresent  Omniscient  Hair Manipulation  \
0  ...            False        False       False              False   
1  ...            False        False       False              False   

   Nova Force  Odin Force  Phoenix Force  Intuitive aptitude  Melting  \
0       False       False          False               False    False   
1       False       False          False               False    False   

   Changing Armor  
0           False  
1           False  

[2 rows x 168 columns]

In [30]:
# check info for merged_powers before merging
merged_powers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 667 entries, 0 to 666
Columns: 168 entries, Hero to Changing Armor
dtypes: bool(167), object(1)
memory usage: 114.1+ KB


In [32]:
# check info for info before merging
dfsh.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 463 entries, 0 to 462
Data columns (total 10 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Publisher    463 non-null    object 
 1   Hero         463 non-null    object 
 2   Gender       463 non-null    object 
 3   Race         463 non-null    object 
 4   Alignment    463 non-null    object 
 5   Hair color   463 non-null    object 
 6   Eye color    463 non-null    object 
 7   Skin color   463 non-null    object 
 8   Height (cm)  463 non-null    float64
 9   Weight (kg)  463 non-null    float64
dtypes: float64(2), object(8)
memory usage: 36.3+ KB


In [34]:
# combine two dataframes on info['Hero'] and merged_powers['Hero']
hero_df = pd.merge(dfsh, 
                   merged_powers, 
                   on = 'Hero')

# check
hero_df.head()

Publisher           Hero Gender               Race Alignment  \
0      Marvel Comics         A-Bomb   Male              Human      good   
1  Dark Horse Comics     Abe Sapien   Male      Icthyo Sapien      good   
2          DC Comics       Abin Sur   Male            Ungaran      good   
3      Marvel Comics    Abomination   Male  Human / Radiation       bad   
4      Marvel Comics  Absorbing Man   Male              Human       bad   

  Hair color Eye color Skin color  Height (cm)  Weight (kg)  ...  \
0    No Hair    yellow    Unknown        203.0        441.0  ...   
1    No Hair      blue       blue        191.0         65.0  ...   
2    No Hair      blue        red        185.0         90.0  ...   
3    No Hair     green    Unknown        203.0        441.0  ...   
4    No Hair      blue    Unknown        193.0        122.0  ...   

   Weather Control  Omnipresent  Omniscient  Hair Manipulation  Nova Force  \
0            False        False       False              False       False   
1            False        False       False              False       False   
2            False        False       False              False       False   
3            False        False       False              False       False   
4            False        False       False              False       False   

   Odin Force  Phoenix Force  Intuitive aptitude  Melting  Changing Armor  
0       False          False               False    False           False  
1       False          False               False    False           False  
2       False          False               False    False           False  
3       False          False               False    False           False  
4       False          False               False    False           False  

[5 rows x 177 columns]

In [35]:
# check
hero_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 463 entries, 0 to 462
Columns: 177 entries, Publisher to Changing Armor
dtypes: bool(167), float64(2), object(8)
memory usage: 115.3+ KB


# II. Use your combined DataFrame to answer the following questions.

Compare the average weight of super powers who have Super Speed to those who do not.
What is the average height of heroes for each publisher?

In [36]:
hero_df.groupby('Super Speed')['Weight (kg)'].mean().round(2)

Super Speed
False    101.77
True     129.40
Name: Weight (kg), dtype: float64

From the information gathered by the dataset, the Super Hero who has Super Speed weighs an average of 129.4 KG.
The Super Hero who does not have Super Speed weighs an average of 101.77 KG.
An interesting further break down would be to compare the genders of the Super Heroes, then their ages.

In [37]:
hero_df.groupby('Publisher')['Height (cm)'].mean().round(2)

Publisher
DC Comics            181.92
Dark Horse Comics    176.91
George Lucas         159.60
Image Comics         211.00
Marvel Comics        191.55
Shueisha             171.50
Star Trek            181.50
Team Epic TV         180.75
Unknown              178.00
Name: Height (cm), dtype: float64

From the information dataset, the publisher with the smallest heroes is George Lucas.  This makes sense if you look at his Star Wars comics.  Several of his heroes are from an alien race, like Yoda or the Ewoks, who are drastically under 140 cm.  The dataset also shows that Image Comics has the biggest Super Heroes. who are and averages of 211 cm.  The rest of the publishers all into this ascending order: Shueisha, Dark Horse, Unknown, Team Epic TV, Star Trek, DC Comics, and Marvel Comics.
